src/frequenz/client/dispatch/_client.py

```diff
@@ -0,0 +1,234 @@
+            end_to: optional, filter by end_time < end_to.
+            is_active: optional, filter by is_active status.
+            is_dry_run: optional, filter by is_dry_run status.
+
+        Returns:
+            DispatchList: List of dispatches.
+        """
+        time_interval = None
+
+        def to_timestamp(dt: datetime | None) -> Timestamp | None:
+            if dt is None:
+                return None
+
+            ts = Timestamp()
+            ts.FromDatetime(dt)
+            return ts
+
+        if start_from or start_to or end_from or end_to:
+            time_interval = PBTimeIntervalFilter(
+                start_from=to_timestamp(start_from),
+                start_to=to_timestamp(start_to),
+                end_from=to_timestamp(end_from),
+                end_to=to_timestamp(end_to),
+            )
+
+        selectors = []
+
+        for selector in component_selectors if component_selectors else []:
+            selectors.append(_component_selector_to_protobuf(selector))
+
+        filters = DispatchFilter(
+            selectors=selectors,
+            time_interval=time_interval,
+            is_active=is_active,
+            is_dry_run=is_dry_run,
+        )
+        request = DispatchListRequest(microgrid_id=microgrid_id, filter=filters)
+
+        response: Awaitable[DispatchList]
+        response = self._stub.ListMicrogridDispatches(request)  # type: ignore
+        return list(map(Dispatch.from_protobuf, (await response).dispatches))
+
+    async def create(
+        self,
+        microgrid_id: int,
+        _type: str,
+        start_time: datetime,
+        duration: timedelta,
+        selector: ComponentSelector,
+        is_active: bool,
+        is_dry_run: bool,
+        payload: dict[str, Any],
+        recurrence: RecurrenceRule,
+    ) -> None:
+        """Create a dispatch.
+
+        Args:
+            microgrid_id: The microgrid_id to create the dispatch for.
+            _type: User defined string to identify the dispatch type.
+            start_time: The start time of the dispatch.
+            duration: The duration of the dispatch.
+            selector: The component selector for the dispatch.
+            is_active: The is_active status of the dispatch.
+            is_dry_run: The is_dry_run status of the dispatch.
+            payload: The payload of the dispatch.
+            recurrence: The recurrence rule of the dispatch.
+        """
+        request = DispatchCreateRequest(
+            microgrid_id=microgrid_id,
+            type=_type,
+            start_time=start_time,
+            duration=duration,
+            selector=selector,
+            is_active=is_active,
+            is_dry_run=is_dry_run,
+            payload=payload,
+            recurrence=recurrence,
+        ).to_protobuf()
+
+        await self._stub.CreateMicrogridDispatch(request)  # type: ignore
+
+    async def update(
+        self,
+        dispatch_id: int,
+        new_fields: dict[str, Any],
```

Comment:
```suggestion
        **new_fields: Any,
```

But I would actually list explicitly everything that can be updated here, otherwise we lose type safety and also we allow introducing bugs because of typos (`update(start_tim=...)` won't be caught by `mypy`) and the code is quite convoluted with the huge `match` statement. Actually now that I see it the `**new_fields: Any` won't work because you use `.` in keys.

For this PR I'm fine with this but I think we should definitely do it more static and type checked in the future.

Reply:
>  the code is quite convoluted with the huge match statement

It would look similar with any other solution as I have to take care / convert every (type of ) member individually.. if it's not a `match` statement it would be a `if para: ...` instead for each parameter..

Reply:
Yeah, is true that you will still check which ones are set to be updated and which are not one by one, but you could at least group `RecurrenceRuleUpdate` in wrapper, so that can be omitted as a group if nothing has to be changed for example, but the main point here for me is still type safety and being able to get errors at *`mypy`time* :D

Reply:
I though about other options, using `TypedDict` for example, to avoid the bloat, but all have their issues, I guess this is something we might need to think more about, so I would leave it as in this PR and create an issue about it. I think how we do updates should also be consistent among all API clients, so is something we might want to discuss with the other API teams.

Reply:
A data class with every type optional as a parameter could be a way.

Reply:
Yeah, we would need to use a sentinel to indicate if a value was set or not (because you might need to update something from a value to `None`, so `None` can't mean "this attribute is not updated") but that still requires you to check one by one every attribute to see if it was overridden or not. `dict`s are the only ones that have the nice property of encoding the absence of a field natively, so you can just iterate over the existing keys to set the attributes in the message.

Reply:
I could also use the top suggestion and use `_` instead of `.` as path separator..

Reply:
For what exactly?

Reply:
For making `**new_fields: Any` possible